MapboxVisionTests/VisionManagerTests.swift

```diff
@@ -0,0 +1,161 @@
+            self?.visionManager = nil
+        }
+    }
+
+    func test_NoAssertOnVisionManagerDeallocAfterDestroy() {
+        // Given
+        // object of VisionManager
+        // When
+        // we deallocate it with destroy
+        // Then
+        // Shouldn't be assertion failure
+        expectNoAssertionFailure { [weak self] in
+            self?.visionManager?.destroy()
+            self?.visionManager = nil
+        }
+    }
+
+    func test_RecordingPermissions() {
+        // Given
+        // VisionManager with default country
+
+        guard let visionManager = self.visionManager, let recorder = dependencies?.recorder as? MockSessionRecorder else {
+            XCTAssert(false)
+            return
+        }
+
+        // When
+        // We set countries
+
+        visionManager.onCountryUpdated(.USA)
+        visionManager.onCountryUpdated(.UK)
+        visionManager.onCountryUpdated(.china)
+        visionManager.onCountryUpdated(.other)
+        visionManager.onCountryUpdated(.unknown)
+
+        // Then
+        // we expect that visionManager will call SessionRecorder's method `startInternal` for all countries but China and that there
+        // will no be any other calls to SessionRecorder. For China we expect `stop` method to be called
+
+        let expectedActions: [MockSessionRecorder.Action] = [
+            .startInternal, // USA
+            .startInternal, // UK
+            .stop,          // China
+            .startInternal, // other
+            .startInternal  // unknown
+        ]
+
+        XCTAssert(recorder.actionsLog.elementsEqual(expectedActions))
+    }
+
+    func test_RecordingPermisionsWhileExternalRecord() {
+        // Given
+        // VisionManager with default country which is in external recording mode
+
+        guard let visionManager = self.visionManager, let recorder = dependencies?.recorder as? MockSessionRecorder else {
+            XCTAssert(false)
+            return
+        }
+
+        visionManager.start()
+        try? visionManager.startRecording(to: "")
+
+        // When
+        // We set countries
+
+        visionManager.onCountryUpdated(.USA)
+        visionManager.onCountryUpdated(.UK)
+        visionManager.onCountryUpdated(.china)
+        visionManager.onCountryUpdated(.other)
+        visionManager.onCountryUpdated(.unknown)
+
+        // Then
+        // we expect that visionManager won't do anything
+
+        let expectedActions: [MockSessionRecorder.Action] = [
+            .startInternal,
+            .stop,
+            .startExternal(withPath: ""),
+            .startInternal,
+            .startInternal,
+            .stop,
+            .startInternal,
+            .startInternal
+        ]
+
+        XCTAssert(recorder.actionsLog.elementsEqual(expectedActions))
+
+        visionManager.stopRecording()
+        visionManager.stop()
+    }
+}
+
+private extension XCTestCase {
+    func expectAssertionFailure(testcase: @escaping () -> Void) {
```

Comment:
WDYT about making the signature of this method as close to XCTest as possible (with message and stuff)? It also may be top-level function.

Reply:
done